### PR TITLE
ENH: Adding new test for GDCM ImageIO

### PR DESCRIPTION
**Pull request by blowekamp**
## PR Checklist

- [X] :no_entry_sign: [Makes breaking changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes)
- [X] :no_entry_sign: [Makes design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes)
- [X] :no_entry_sign: Adds the License notice to new files.
- [X] :no_entry_sign: Adds Python wrapping.
- [X] :no_entry_sign: Adds tests and baseline comparison (quantitative).
- [X] :no_entry_sign: [Adds test data](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Documentation/UploadBinaryData.md).
- [X] :no_entry_sign: Adds Examples to [ITKExamples](https://github.com/InsightSoftwareConsortium/ITKExamples)
- [X] :no_entry_sign: Adds Documentation.

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

File: Modules/IO/GDCM/test/CMakeLists.txt [hunk 1]
```diff
@@ -163,3 +164,12 @@ itk_add_test(NAME itkGDCMImageIONoPreambleTest
   COMMAND ITKIOGDCMTestDriver itkGDCMImageIONoPreambleTest
   DATA{Input/NoPreambleDicomTest.dcm}
   )
+
+itk_add_test(NAME itkGDCMImageReadWriteTest_RGB
```
Comment (dzenanz) [inline Modules/IO/GDCM/test/CMakeLists.txt]: Perhaps add one more test here, to establish the pattern of how to add more tests?

**Comment by dzenanz** [comment]
Why are Azure Pipelines not running on this PR?

**Comment by blowekamp** [comment]
This PR is separate from the fix in #1309, additional cases can be tested when combined with PR #1309 . We are [discussing creating suitable small test images](https://github.com/InsightSoftwareConsortium/ITK/pull/1309#issuecomment-544999523). To move forward, one PR needs to be merged OR they need to be combined. There seems to be a lot of flux with the other PR, so I am reluctant to combine them.

I tries adding a test for "JPEG2000_YBR_RCT--1.dcm" with the additions of PR and I get the following error:
```
ITKIOGDCMTestDriver: /home/blowekamp/src/ITK/Modules/ThirdParty/GDCM/src/gdcm/Source/MediaStorageAndFileFormat/gdcmJPEG2000Codec.cxx:789: std::pair<char*, long unsigned int> gdcm::JPEG2000Codec::DecodeByStreamsCommon(char*, size_t): Assertion `!mct' failed.
1/1 Test #904: itkGDCMImageReadWriteTest_JPEG2000_YBR_ICT ...***Exception: Child aborted  0.20 sec
```

I suggest that this PR is merged then PR #1309 is rebased onto this where some tests can be added.

**Comment by issakomi** [comment]
The only image with VECTOR type from that [collection](https://drive.google.com/drive/folders/1hhPDBf_J1i0qmtYlsetuIgulEU2Ui_J7?usp=sharing) will be CMYK (if #1309 will be pushed). It is ca. 500 KB. So i have to create small CMYK (or may be HSV or ARGB) image?

Is it correct?

Alternatively YBR_FULL and YBR_FULL_422 can be type VECTOR if _SetReadYBRtoRGB(false)_ will be set. If #1309 will be dropped there will be no VECTOR type at all. Actually (and if #1310 will be preferred) the logic is
```
if (number of components == 1) SCALAR
else RGB
```
After #1309 VECTOR could be

- CMYK, HSV, ARGB (all retired and very rare)
- YBR_FULL, YBR_FULL_422 with _SetReadYBRtoRGB(false)_

**Comment by issakomi** [comment]
@blowekamp 

```
TKIOGDCMTestDriver: /home/blowekamp/src/ITK/Modules/ThirdParty/GDCM/src/gdcm/Source/MediaStorageAndFileFormat/gdcmJPEG2000Codec.cxx:789: std::pair<char*, long unsigned int> gdcm::JPEG2000Codec::DecodeByStreamsCommon(char*, size_t): Assertion `!mct' failed.
1/1 Test #904: itkGDCMImageReadWriteTest_JPEG2000_YBR_ICT ...***Exception: Child aborted  0.20 sec
```
I know, in release mode everything is OK with the image, it is D. Clunie's image, IMHO @malaterre  knows more about assert

**Comment by blowekamp** [comment]
> The only image with VECTOR type from that [collection](https://drive.google.com/drive/folders/1hhPDBf_J1i0qmtYlsetuIgulEU2Ui_J7?usp=sharing) will be CMYK (if #1309 will be pushed). It is ca. 500 KB. So i have to create small CMYK (or may be HSV or ARGB) image?

Those exotic cases are lower priority. But the test image size's should be closer to 50k. Can they be relatively "easily" just cropped? If this is a difficult task we can discuss how many larger image should be include. Ideally well have a bunch of small images to test these cases.





> Is it correct?
> 
> Alternatively YBR_FULL and YBR_FULL_422 can be type VECTOR if _SetReadYBRtoRGB(false)_ will be set. If #1309 will be dropped there will be no VECTOR type at all. Actually (and if #1310 will be preferred) the logic is
> 
> ```
> if (number of components == 1) SCALAR
> else RGB
> ```
> 
> After #1309 VECTOR type is
> 
> * CMYK, HSV, ARGB (all retired and very rare)
> * YBR_FULL, YBR_FULL_422 with _SetReadYBRtoRGB(false)_

We are working towards getting #1309 merged, but the features need to be tested. Any idea why I'm getting the failed assertion reported above?

**Comment by issakomi** [comment]
> Any idea why I'm getting the failed assertion reported above?

No. But is not related to #1309 or #1310.

**Comment by dzenanz** [comment]
Let's wait for CI, merge this PR, rebase #1309 on top of it, and continue discussion there?

**Comment by issakomi** [comment]
JPEG2000 have relatively often asserts, sometimes also like is this [discussion](https://discourse.itk.org/t/photometricinterpretation/1837). So far in release mode always no problems.
But we didn't touch JPEG2000, they are YBR_ICT/YBR_RCT, but always open in RGB, so no post-processing like with YBR_FULL/_422 was required.

**Comment by issakomi** [comment]
BTW, the test assumes always 2D image, i am not sure what kind of images will be taken for tests, but many from that collection are ultrasound multi-frame. Just FYI.

**Comment by issakomi** [comment]
Here is small 32 x 32 (5 KB) CMYK (simple, single color) to test 4 components (if required) [raw_CMYK--2.dcm](https://drive.google.com/file/d/1pDnT1fEhgbm0FPBvDFUkgnVlzafoxFCR/view?usp=sharing)

**Comment by malaterre** [comment]
@blowekamp The assert() issue has been fixed in GDCM release branch. Sorry about that, I totally missed `YBR_ICT` at that time. I'll submit a PR to update GDCM to latest release, which should also fix a couple of other issues.

@issakomi I am strongly opposed to having a CMYK sample dataset in the regular test suite of ITK. I have never seen a real-life sample of such beast. `CMYK` has been retired since (I would say) at least 2003. Let's focus on actual real-life DICOM color sample dataset.

ref:
* http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.6.3.html#sect_C.7.6.3.1.2

**Comment by issakomi** [comment]
> @issakomi I am strongly opposed to having a CMYK sample dataset in the regular test suite of ITK.

I thought the test is about pixel type, it is retired, but interesting type - vector with 4 components. But i don't insist at all.

**Comment by blowekamp** [comment]
> I thought the test is about pixel type,...

That is likely not the correct phrase... The test is rather generic as it's to just read, write and verify the results are as expected. Perhaps it is to test different pixel encodings too?

> BTW, the test assumes always 2D image, i am not sure what kind of images will be taken for tests, but many from that collection are ultrasound multi-frame. Just FYI.

Only some many things to check at one time. Let us get a catalog of image/pixel "encodings" to test of simple 2D images.

On the todo list after merging and integration with the other PR, is to add a one ( or more ) command like flags for options like `ReadYBRtoRGB` or the other palette to RGB option if ever supported.

**Comment by issakomi** [comment]
> > I thought the test is about pixel type,...
> 
> That is likely not the correct phrase... supported.

The name of the PR is ocasionally _ENH: Adding new test for GDCM ImageIO for different pixel types_. :) Anyway, deleted old collection. Hope new one with generated small images is OK.

**Comment by blowekamp** [comment]
Looks like the AZP CI is passing. Can we merge?

**Pull request merged**